Python: TableScan Plan files API implementation without residual evaluation

dhruv-pratap opened the pull request:
Taking a first dig at TableScan Plan Files API. The implementation at present evaluates partition filters but does not evaluate row-level filters at the moment as it requires Projections.

The API is exposed by the CLI tool and lists out all the data files that would need to be scanned in order to perform the scan plan.

<img width="1912" alt="image" src="https://user-images.githubusercontent.com/11399509/198372507-7f4a9b8a-bd3d-44d8-a534-234bacb0bba0.png">

High-level changes done as part of this PR:
1. Add `new_scan()` API to the `Table` interface.
2. Add `TableScan` interface with just `plan_files` abstract method.
3. Add `FileScanTask` model that represents a scan task over the existing `DataFile` model.
4. Add `DataTableScan` class that extends `TableScan` and implements `plan_files` API to return a collection of `FileScanTasks` using the existing `_ManifestEvalVisitor` to perform partition pruning.

Fokko commented on python/pyiceberg/cli/console.py:
```diff
@@ -361,3 +361,24 @@ def table(ctx: Context, identifier: str, property_name: str):  # noqa: F811
+@click.argument("identifier")
+@click.pass_context
+@catch_exception()
+def scan_table(ctx: Context, identifier: str):
```
I like that we can easily visualize the files. I think we should merge this one with the files command since they are so similar. WDYT?

```
➜  python git:(master) ✗ pyiceberg --verbose true --catalog fokko files hive_test.h1
Snapshots: fokko.hive_test.h1
└── Snapshot 4362649644256929116, schema 0: 
    s3://bucket/41246f47-e1bc-465e-89b4-6d039a9d55dc/67ad77ac-d90b-4472-b1e9-672d91a1ac41/metadata/snap-4362649644256929116-1-63e88d4d-1b30-4cae-b7a0-f53b0b54758a.avro
    ├── Manifest: s3://bucket/41246f47-e1bc-465e-89b4-6d039a9d55dc/67ad77ac-d90b-4472-b1e9-672d91a1ac41/metadata/63e88d4d-1b30-4cae-b7a0-f53b0b54758a-m0.avro
    │   └── Datafile: 
    │       s3://bucket/41246f47-e1bc-465e-89b4-6d039a9d55dc/67ad77ac-d90b-4472-b1e9-672d91a1ac41/data/0bfca5ff//00000-0-dweeks_20220901095219_f181f97a-43ce-4aa6-8cec-6956d83b7a42-job_lo
    │       cal2002913683_0002-00001.parquet
    └── Manifest: s3://bucket/41246f47-e1bc-465e-89b4-6d039a9d55dc/67ad77ac-d90b-4472-b1e9-672d91a1ac41/metadata/508fd9ac-520f-47d8-900f-00f1dda051ac-m0.avro
        └── Datafile: s3://bucket/41246f47-e1bc-465e-89b4-6d039a9d55dc/67ad77ac-d90b-4472-b1e9-672d91a1ac41/data/3d030bcf/00000-0-44897130-c494-4dca-a164-e874c1e37c36-00001.parquet
```

dhruv-pratap replied:
Not sure I follow you @Fokko . Do you mean make `scan` a sub-command of `files` command? Or just have `files` command use the `table.new_scan().plan_files()` API underneath?

If it's the later, I like that.

Fokko replied:
I think the later one is more flexible, but then we should remove the files because they have so much overlap

dhruv-pratap replied:
The only difference being `files` at present lists all snapshots. The table scan `plan_files()` will only list the snapshot that is being used for the scan.

dhruv-pratap replied:
Maybe we can make files with three options:
- `files --snapshot=all` - The current files behavior
- `files --snapshot=current` - Files under the current snapshot
- `files --snapshot=<snapshot_id>` - Files under the snapshot_id specified

Fokko commented on python/pyiceberg/expressions/visitors.py:
```diff
@@ -613,6 +613,8 @@ def visit_or(self, left_result: bool, right_result: bool) -> bool:
-    partition_schema = Schema(*partition_spec.partition_type(schema))
+    partition_type = partition_spec.partition_type(schema)
+    partition_schema = Schema(*partition_type.fields)
+    partition_filter = partition_filter if partition_filter is not None else AlwaysTrue()
```
Technically this should not happen. The `partition_filter` is non-optional. A shorter way of writing this is:
```suggestion
    partition_filter = partition_filter or AlwaysTrue()
```

Fokko commented on python/pyiceberg/table/__init__.py:
```diff
@@ -60,33 +60,40 @@ def sort_order(self) -> SortOrder:
         """Get the snapshot history of this table."""
         return self.metadata.snapshot_log
+
+    def new_scan(self, io: FileIO, snapshot_id: int | None = None, expression: BooleanExpression | None = None):
```
What do you think of setting it to AlwaysTrue by default?
```suggestion
    def new_scan(self, io: FileIO, snapshot_id: int | None = None, expression: BooleanExpression = AlwaysTrue()):
```

Fokko commented on python/pyiceberg/table/__init__.py:
```diff
@@ -60,33 +60,40 @@ def sort_order(self) -> SortOrder:
+        """Create a new scan for this table."""
+        from pyiceberg.table.scan import DataTableScan
+
+        snapshot = self.snapshot_by_id(snapshot_id) if snapshot_id is not None else None
```
What do you think of setting the current `snapshot_id` if it hasn't been passed in explicitly?
```python
snapshot_id = snapshot_id or self.metadata.current_snapshot_id
snapshot = self.snapshot_by_id(snapshot_id)
```

dhruv-pratap replied:
That does make sense. Will amend.

dhruv-pratap replied:
Amended in a recent commit.

Fokko commented on python/pyiceberg/table/scan.py:
```diff
@@ -0,0 +1,115 @@
+
+    io: FileIO
+
+    def __init__(self, io: FileIO, table: Table, snapshot: Snapshot | None = None, expression: BooleanExpression | None = None):
```
How about passing in the schema and partition-spec, rather than pulling in the whole table object?

Fokko reviewed (commented):
Thanks for working on this @dhruv-pratap I have some suggestions, let me know what you think!

dhruv-pratap commented on python/pyiceberg/cli/output.py:
```diff
@@ -14,9 +14,11 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+from __future__ import annotations
```
I like Optional[str] too, but I think for some reason when I run `make lint` locally it turns it into the later format i.e. `str | None ` Do we need to tweak the linter config to enforce one format over the other?

Fokko replied:
This is because the import statement. PyUpgrade will automatically rewrite the imports: https://github.com/asottile/pyupgrade#pep-604-typing-rewrites We can change this in another PR.

dhruv-pratap replied:
I've reverted to not use `from __future__ import annotations` in a recent commit.

Fokko replied:
It should have been fixed now https://github.com/apache/iceberg/pull/6114 is in. Delayed loading of the annotations (that's what the import does), can be helpful sometime :)

dhruv-pratap commented:
Added tests for table scan with partition filters. Also, added logic to copy and rewrite a table's metadata along with its data. This logic can be used to stage test table in a temporary directory to perform scan operations on it.  This seems to be a necessity in absence of write-path APIs to stage test table for testing read-path APIs.

Happy to listen to suggestions if there are any alternatives to test these.

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -38,7 +41,7 @@ def exception(self, ex: Exception) -> None:
 
     @abstractmethod
-    def identifiers(self, identifiers: List[Identifier]) -> None:
+    def identifiers(self, identifiers: list[Identifier]) -> None:
```
```suggestion
    def identifiers(self, identifiers: List[Identifier]) -> None:
```

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -49,6 +52,10 @@ def describe_table(self, table: Table) -> None:
         ...
 
+    @abstractmethod
+    def scan_plan_files(self, plan_files: Iterable[FileScanTask], snapshot_id: int | None = None) -> None:
```
```suggestion
    def scan_plan_files(self, plan_files: Iterable[FileScanTask], snapshot_id: Optional[int] = None) -> None:
```

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -66,7 +73,7 @@ def spec(self, spec: PartitionSpec) -> None:
 
     @abstractmethod
-    def uuid(self, uuid: Optional[UUID]) -> None:
+    def uuid(self, uuid: UUID | None) -> None:
```
```suggestion
    def uuid(self, uuid: Optional[UUID]) -> None:
```

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -88,7 +95,7 @@ def exception(self, ex: Exception) -> None:
             Console(stderr=True).print(ex)
 
-    def identifiers(self, identifiers: List[Identifier]) -> None:
+    def identifiers(self, identifiers: list[Identifier]) -> None:
```
```suggestion
    def identifiers(self, identifiers: List[Identifier]) -> None:
```

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -146,6 +153,19 @@ def files(self, table: Table, io: FileIO, history: bool) -> None:
+
+        manifest_dict = {}
+        for file in plan_files:
+            if file.manifest.manifest_path not in manifest_dict:
```
Nit: Should we move `file.manifest.manifest_path` in a local variable, since we use it three times?

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -164,7 +184,7 @@ def schema(self, schema: Schema) -> None:
         Console().print(str(spec))
 
-    def uuid(self, uuid: Optional[UUID]) -> None:
+    def uuid(self, uuid: UUID | None) -> None:
```
```suggestion
    def uuid(self, uuid: Optional[UUID]) -> None:
```

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -182,7 +202,7 @@ def _out(self, d: Any) -> None:
         self._out({"type": ex.__class__.__name__, "message": str(ex)})
 
-    def identifiers(self, identifiers: List[Identifier]) -> None:
+    def identifiers(self, identifiers: list[Identifier]) -> None:
```
```suggestion
    def identifiers(self, identifiers: List[Identifier]) -> None:
```

Fokko commented on python/pyiceberg/cli/output.py:
```diff
@@ -200,8 +220,11 @@ def schema(self, schema: Schema) -> None:
         print(spec.json())
 
-    def uuid(self, uuid: Optional[UUID]) -> None:
+    def uuid(self, uuid: UUID | None) -> None:
```
```suggestion
    def uuid(self, uuid: Optional[UUID]) -> None:
```

Fokko commented on python/pyiceberg/expressions/visitors.py:
```diff
@@ -436,8 +436,10 @@ def __init__(self, partition_struct_schema: Schema, partition_filter: BooleanExp
         self.partition_filter = bind(partition_struct_schema, rewrite_not(partition_filter), case_sensitive)
 
     def eval(self, manifest: ManifestFile) -> bool:
+        print(f"Evaluating ManifestFile = {manifest} with partition_filter={self.partition_filter}")
```
I think this is still around from the debugging? Maybe convert it into a log statement?

Fokko commented on python/pyiceberg/expressions/visitors.py:
```diff
@@ -517,7 +519,7 @@ def visit_equal(self, term: BoundTerm, literal: Literal[Any]) -> bool:
         field = self.partition_fields[pos]
 
-        if field.lower_bound is None:
+        if field.lower_bound is None or field.upper_bound is None:
```
Great catch! Thanks!

rdblue replied:
If one is `None` then both should be, so this shouldn't have any effect. But it's fine to update this.

Fokko replied:
This is required for the linter since we're the variable below.

Fokko commented on python/pyiceberg/expressions/visitors.py:
```diff
@@ -613,6 +615,9 @@ def visit_or(self, left_result: bool, right_result: bool) -> bool:
-    partition_schema = Schema(*partition_spec.partition_type(schema))
+    partition_type = partition_spec.partition_type(schema)
+    partition_schema = Schema(*partition_type.fields)
+    partition_filter = partition_filter or AlwaysTrue()
```
Since `partition_filter` cannot be `None` according to the type, this statement doesn't have any effect

Fokko commented on python/pyiceberg/table/__init__.py:
```diff
@@ -90,3 +92,15 @@ def snapshot_by_name(self, name: str) -> Optional[Snapshot]:
         """Get the snapshot history of this table."""
         return self.metadata.snapshot_log
+
+    def new_scan(self, io: FileIO, snapshot_id: Optional[int] = None, expression: Optional[BooleanExpression] = ALWAYS_TRUE):
```
```suggestion
    def new_scan(self, io: FileIO, snapshot_id: Optional[int] = None, expression:BooleanExpression = ALWAYS_TRUE):
```
Since we default to `AlwaysTrue`, I think we can set this as required

Fokko commented on python/pyiceberg/table/__init__.py:
```diff
@@ -90,3 +92,15 @@ def snapshot_by_name(self, name: str) -> Optional[Snapshot]:
+        """Create a new scan for this table."""
+        from pyiceberg.table.scan import DataTableScan
+
+        if not (use_snapshot := snapshot_id or self.metadata.current_snapshot_id):
```
The walrus operator doesn't add any value here, I would suggest removing it
```suggestion
        if not snapshot_id and not self.metadata.current_snapshot_id:
```

Fokko commented on python/pyiceberg/table/__init__.py:
```diff
@@ -90,3 +92,15 @@ def snapshot_by_name(self, name: str) -> Optional[Snapshot]:
+        if not (use_snapshot := snapshot_id or self.metadata.current_snapshot_id):
+            raise ValueError("Unable to resolve a snapshot to use for this scan.")
+
+        if not (snapshot := self.snapshot_by_id(use_snapshot)):
```
Same thing here, the walrus operator doesn't add anything here
```suggestion
        if not self.snapshot_by_id(use_snapshot):
```

Fokko commented on python/pyiceberg/table/scan.py:
```diff
@@ -0,0 +1,103 @@
+    snapshot: Snapshot
+    expression: BooleanExpression
+
+    def __init__(self, table: Table, snapshot: Optional[Snapshot] = None, expression: Optional[BooleanExpression] = ALWAYS_TRUE):
```
Since we provide a default here, maybe we should just make it mandatory
```suggestion
    def __init__(self, table: Table, snapshot: Optional[Snapshot] = None, expression: BooleanExpression = ALWAYS_TRUE):
```

Fokko commented on python/pyiceberg/table/scan.py:
```diff
@@ -0,0 +1,103 @@
+        self, io: FileIO, table: Table, snapshot: Optional[Snapshot] = None, expression: Optional[BooleanExpression] = ALWAYS_TRUE
+    ):
+        self.io = io
+        super().__init__(table, snapshot, expression)
```
Nit: Convention is to first call the super, and then do the rest

Fokko commented on python/tests/table/test_scan.py:
```diff
@@ -0,0 +1,177 @@
+
+
+@pytest.fixture
+def temperatures_table() -> Table:
```
I'm hesitant to keep actual binaries in the repository. We already generate the manifest list and manifest itself: https://github.com/apache/iceberg/blob/b215c4861a0c967f8d2a2b0444697d7877d33992/python/tests/conftest.py#L952-L967 How would you feel about generating some Parquet files too?

rdblue replied:
Yes, this should generate a sample table rather than putting binaries in the repo.

Initially, we can build all of this based on just Avro and add Parquet tests later.

dhruv-pratap replied:
Alright, let me take a look at that example, reverse engineer the binaries, and remove it from the repo.

rdblue commented on python/pyiceberg/expressions/visitors.py:
```diff
@@ -526,7 +528,7 @@ def visit_equal(self, term: BoundTerm, literal: Literal[Any]) -> bool:
             return ROWS_CANNOT_MATCH
 
-        upper = _from_byte_buffer(term.ref().field.field_type, field.lower_bound)
+        upper = _from_byte_buffer(term.ref().field.field_type, field.upper_bound)
```
@Fokko, can you check why tests didn't catch this case? We should ensure this is tested properly and separate the fix into its own PR.

rdblue commented on python/pyiceberg/table/scan.py:
```diff
@@ -0,0 +1,103 @@
+        return self.data_file.file_size_in_bytes
+
+
+class TableScan(ABC):
```
Why does this API not match the Java API? Is there something more pythonic about this?

dhruv-pratap replied:
Just wanted to build the API incrementally and add things as and when needed, otherwise it becomes a bloated PR.

rdblue commented on python/pyiceberg/table/scan.py:
```diff
@@ -0,0 +1,103 @@
+from pyiceberg.table import PartitionSpec, Snapshot, Table
+from pyiceberg.utils.iceberg_base_model import IcebergBaseModel
+
+ALWAYS_TRUE = AlwaysTrue()
```
I'd prefer not adding these constants everywhere. Just use `AlwaysTrue()` in place of this. It's a singleton.

rdblue commented on python/pyiceberg/table/scan.py:
```diff
@@ -0,0 +1,103 @@
+            raise ValueError("Unable to resolve to a Snapshot to use for the table scan.")
+
+    @abstractmethod
+    def plan_files(self) -> Iterable[FileScanTask]:
```
In Java, there are several different types of tasks. I'd recommend using the same pattern and not requiring this to be `FileScanTask`. Introduce an abstract `ScanTask` and return an iterable of that instead.

rdblue commented:
@Fokko, @dhruv-pratap, I posted an alternative scan API in a draft as #6131. Please take a look. That behaves like this one and allows you to specify optional arguments when creating a scan, but also allows the same syntax that Java uses.

dhruv-pratap commented:
In retrospect, I think this is becoming too large of a PR and would benefit from breaking down into smaller tasks. I'm going to go ahead and close this PR and if you guys are onboard I can create issues for the below:

1. Add `ScanTask` interface
2. Add `FileScanTask` class that extends `ScanTask` interface
3. Fix ManifestEvalVisitor eq #6117 
4. Add MetricsEvalVisitor that evaluates an expression on a DataFile to test whether rows in the file may match.
5. Add ResidualEvalVisitor that evaluates the residuals for an expression the partitions in the given PartitionSpec.
6. Add `TableScan` interface. @rdblue just picked this up in #6131 
7. Add `DataTableScan` class that extends implements `plan_files()` API
8. Expose plan_files API as scan subcommand via CLI.

CC: @Fokko @rdblue @samredai

Fokko commented:
Hey @dhruv-pratap that makes a lot of sense. Maybe we should create issues on the list you mentioned above, to make sure that we're aligned on who's working on what. Smaller PRs make it much easier to get stuff merged.

rdblue commented:
@dhruv-pratap, we've been working on the list lately and I think the remaining items are: (4) add `MetricsEvalVisitor` to prune by column stats, (5) add `ResidualEvalVisitor` to produce residuals (LOW priority), and (8) expose `plan_files` via the CLI.

For the last one, we may just want to add filters to the existing `files` command?

dhruv-pratap commented:
> @dhruv-pratap, we've been working on the list lately and I think the remaining items are: (4) add `MetricsEvalVisitor` to prune by column stats, (5) add `ResidualEvalVisitor` to produce residuals (LOW priority), and (8) expose `plan_files` via the CLI.
> 
> For the last one, we may just want to add filters to the existing `files` command?

@rdblue We discussed this more in [this](https://github.com/apache/iceberg/pull/6069#discussion_r1012044761) thread. We can extend `files` with three options:

* `files --snapshot=all` - The current files behavior
* `files --snapshot=current` - Files under the current snapshot
* `files --snapshot=<snapshot_id>` - Files under the snapshot_id specified

Adding `expression` option would be a bigger undertaking though, as it would require an expression parser. Not sure if we want to prioritize that at this point for the CLI.

rdblue commented:
> Adding expression option would be a bigger undertaking though, as it would require an expression parser. Not sure if we want to prioritize that at this point for the CLI.

We added an expression parser a few weeks ago, so this would be unblocked. I'm all for updating the CLI with your suggestions.

Fokko commented:
Hey @dhruv-pratap. I'm closing this draft as this has already been implemented. Feel free to open a new PR if you feel something is missing.
